File: LC/Medium/373.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        heap = []
        
        n = min(k, len(nums1))
        m = min(k, len(nums2))
        
        for i in range(n):
            for j in range(m):
                if len(heap) < k:
                    heappush(heap, (-(nums1[i] + nums2[j]), [nums1[i], nums2[j]]))
                else:
                    if nums1[i] + nums2[j] > -heap[0][0]:
                        break
                    heappushpop(heap, (-(nums1[i] + nums2[j]), [nums1[i], nums2[j]]))
            
        
        return [[arr[0], arr[1]] for (_, arr) in heap]

File: LC/C++/Medium/373.cpp
class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        using T = tuple<int, int, int>;
        int n1 = nums1.size(), n2 = nums2.size();

        std::vector<std::vector<int>> ans;
        ans.reserve(k);

        std::priority_queue<T, std::vector<T>, std::greater<T>> pq;
        for (int i = 0; i < std::min(k, n1); ++i) {
            pq.emplace(nums1[i] + nums2[0], i, 0);
        }
        
        while (k-- > 0 && !pq.empty()) {
            auto [currSum, i, j] = pq.top();
            pq.pop();

            ans.push_back({nums1[i], nums2[j]});

            if (j + 1 < n2) {
                pq.emplace(nums1[i] + nums2[j + 1], i, j + 1);
            } 
        }

        return ans;
    }
};